[IDE] Call into `matchWitnessStructure` to check if we should inherit a comment from an overridden declaration

ahoppen opened the pull request:
Instead of duplicating logic whether a witness can match a requirement structurally (like static vs. non-static, matching decl kind etc.) call into the type checker to determine that. 

This required moving `getDocCommentProvidingDecl` and `getCascadingDocComment` from AST to IDE.

Matching the type seemed to work fairly differently between the type checker and for IDE purposes, so we are not sharing that.

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
[IDE] Call into `matchWitnessStructure` to check if we should inherit a comment from an overridden declaration

ahoppen commented:
@swift-ci Please smoke test

hamishknight commented on include/swift/AST/Decl.h:
```diff
@@ -1202,10 +1202,6 @@ class alignas(1 << DeclAlignInBits) Decl : public ASTAllocated<Decl> {
 
   std::optional<unsigned> getSourceOrder() const;
 
-  /// \returns The brief comment attached to this declaration, or the brief
-  /// comment attached to the comment providing decl.
-  StringRef getSemanticBriefComment() const;
-
```
In fact it might be worth defining a request for `getDocCommentProvidingDecl` so we can pull the same trick for that

ahoppen replied:
Moved `getDocCommentProvidingDecl` to be a method on `Decl` as well.

hamishknight commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -499,15 +499,12 @@ checkEffects(AbstractStorageDecl *witness, AbstractStorageDecl *req) {
-    llvm::function_ref<RequirementMatch(bool, ArrayRef<OptionalAdjustment>)>
-        finalize) {
-
+/// Implementation of `matchWitnessStructure` that also sets a few out paramter
```
```suggestion
/// Implementation of `matchWitnessStructure` that also sets a few out parameters
```

hamishknight commented on lib/Sema/TypeChecker.h:
```diff
@@ -487,6 +487,13 @@ void diagnoseDuplicateCaptureVars(CaptureListExpr *expr);
+/// whether it has a matching kind like static vs. non-static, matching decl,
+/// etc. It does not check if the types of the witness is able to satisfy the
+/// requirement.
+/// If there is a matching failure, returns a `true`, otherwise return `false`
+bool witnessStructureMatches(ValueDecl *req, const ValueDecl *witness);
```
This should be the opposite, right? i.e returns `true` on success

ahoppen replied:
Indeed. Thanks for catching.

ahoppen committed:
[Sema] Move `getDocCommentProvidingDecl` and `getCascadingDocComment` from `AST` to `Sema`

This allows use to re-use logic from Sema in those requests.

This commit just moves functions around and does not change any functionality.

ahoppen commented:
@swift-ci Please smoke test

hamishknight commented on lib/Sema/Comment.cpp:
```diff
@@ -0,0 +1,224 @@
+};
+} // end anonymous namespace
+
+const Decl *Decl::getDocCommentProvidingDecl() const {
+  // Search for the first decl we see with a non-empty raw comment.
+  auto finder = CommentProviderFinder<RawComment>(
+      [](const Decl *D) -> std::optional<RawComment> {
+        auto comment = D->getRawComment();
+        if (comment.isEmpty())
+          return std::nullopt;
+        return comment;
+      });
+
+  auto result = finder.findCommentProvider(this);
+  return result ? result->second : nullptr;
+}
+
+StringRef swift::Decl::getSemanticBriefComment() const {
+  if (!canHaveComment())
+    return StringRef();
+
+  auto &eval = getASTContext().evaluator;
+  return evaluateOrDefault(eval, SemanticBriefCommentRequest{this},
+                           StringRef());
+}
```
I probably should have clarified why requests are able to do layering violations like this, it's because the request classes themselves are in libAST, but their evaluation functions are registered at runtime

ahoppen replied:
Ah, I see. That makes me less concerned about the layering as well. I created requests fro both `getSemanticBriefComment` and `getDocCommentProvidingDecl` and moved the code that calls them to `libAST`.

ahoppen commented:
@swift-ci Please smoke test
